src/realm/list.cpp

```diff
@@ -153,10 +167,19 @@ void Lst<T>::distinct(std::vector<size_t>& indices, util::Optional<bool> sort_or
 {
     indices.clear();
     sort(indices, sort_order.value_or(true));
-    auto tree = m_tree.get();
-    auto duplicates = std::unique(indices.begin(), indices.end(), [tree](size_t i1, size_t i2) noexcept {
-        return tree->get(i1) == tree->get(i2);
-    });
+    auto duplicates = indices.end();
+
+    if constexpr (std::is_same_v<T, Mixed>) {
+        duplicates = std::unique(indices.begin(), indices.end(), [this](size_t i1, size_t i2) noexcept {
+            return get(i1) == get(i2);
+        });
+    }
+    else {
+        duplicates = std::unique(indices.begin(), indices.end(), [this](size_t i1, size_t i2) noexcept {
+            return m_tree->get(i1) == m_tree->get(i2);
```

Comment:
I think it is more efficient to capture the tree pointer as in the original implementation.